lib/sync/replication-state.js

```diff
@@ -0,0 +1,279 @@
+    const lengths = states.map((state) => state.length)
+
+    if (!lengths.every((length) => length === local.length)) {
+      return false
+    }
+
+    for (const state of states) {
+        if (state.want > 0) {
+            return false
+        }
+    }
+
+    return true
+  }
+
+  close() {
+    this.removeAllListeners()
+  }
+
+  #createState(options = { remote: false }) {
+    return {
+      remote: options.remote,
+      have: 0,
+      want: 0,
+      length: 0
+    }
+  }
+
+  #setInitialCoreState(core) {
+    const localBitfield = core.core.bitfield
+    const state = this.#createState()
+    state.length = getMaxLength(core)
+
+    for (let i = 0; i < state.length; i++) {
+      if (!localBitfield.get(i)) {
+        state.want++
+      } else {
+        state.have++
+      }
+    }
+
+    this.#setPeerState(core.key, state)
+  }
+
+  #setPeerState(peerPublicKey, state) {
+    const peerId = keyToId(peerPublicKey)
+    this.#state.set(peerId, state)
+    this.emit('state', state)
+    if (this.isSynced()) {
+      this.emit('synced', state)
+    }
+  }
+
+  #updatePeerState(peer) {
+    const localBitfield = this.#core.core.bitfield
+    const remoteBitfield = peer.remoteBitfield
+    const localState = this.#createState()
+    const remoteState = this.#createState({ remote: true })
+
+    const length = getMaxLength(this.#core)
+    let haveLocal = 0
+    let haveRemote = 0
+
+    for (let i = 0; i < length; i++) {
+        const local = localBitfield.get(i)
+        const remote = remoteBitfield.get(i)
+
+        if (local) {
+            haveLocal++
+        }
+
+        if (remote) {
+            haveRemote++
+        }
+    }
+
+    localState.have = haveLocal
+    localState.want = length - haveLocal
+    localState.length = length
+
+    remoteState.have = haveRemote
+    remoteState.want = length - haveRemote
+    remoteState.length = length
+
+
+    this.#setPeerState(this.#core.key, localState)
+    this.#setPeerState(peer.remotePublicKey, remoteState)
+  }
+
+  #getLocalState() {
+    const id = keyToId(this.#core.key)
+    return this.#state.get(id) || this.#createState()
+  }
+
+  async #onPeer(peer) {
+    await this.#core.update()
+
+    this.#core.on('download', (index, byteLength, from) => {
+      if (!from.remotePublicKey.equals(peer.remotePublicKey)) return
+      const state = this.#getLocalState()
+      state.length = getMaxLength(this.#core)
+
+      if (state.have === 0 && state.want === 0) {
+        state.want = state.length
+      }
+
+      state.want--
+      state.have++
+      this.#setPeerState(this.#core.key, state)
+    })
+
+    const peerOnBitfield = peer.onbitfield
+    const peerOnRange = peer.onrange
+
+    peer.onbitfield = (bitfield) => {
+      peerOnBitfield.call(peer, bitfield)
+      this.#updatePeerState(peer)
+    }
+
+    peer.onrange = ({ drop, start, length }) => {
+      peerOnRange.call(peer, { drop, start, length })
+      this.#updatePeerState(peer)
+    }
```

Comment:
Yeah I'm not a fan of it either. I'd like to revise `#updatePeerState` to take the drop, start, and length options and use those instead of looping through the whole bitfield.

So far, though, looping through thousands of records doesn't seem to be a significant issue.

I'd lean toward working on this (and a couple of the other perf suggestions) after setting up a way to do benchmarking.